src/util/bitset.h

```diff
@@ -0,0 +1,463 @@
+    public:
+        constexpr IteratorEnd(const IteratorEnd&) = default;
+    };
+    /** Iterator type returned by begin(), which efficiently iterates all 1 positions. */
+    class Iterator
+    {
+        friend class IntBitSet;
+        I m_val; /**< The original integer's remaining bits. */
+        unsigned m_pos; /** Last reported 1 position (if m_pos != 0). */
+        constexpr Iterator(I val) noexcept : m_val(val), m_pos(0)
+        {
+            if (m_val != 0) m_pos = std::countr_zero(m_val);
+        }
+    public:
+        /** Do not allow external code to construct an Iterator. */
+        Iterator() = delete;
+        // Copying is allowed.
+        constexpr Iterator(const Iterator&) noexcept = default;
+        constexpr Iterator& operator=(const Iterator&) noexcept = default;
+        /** Test whether we are done (can only compare with IteratorEnd). */
+        constexpr friend bool operator==(const Iterator& a, const IteratorEnd&) noexcept
+        {
+            return a.m_val == 0;
+        }
+        /** Progress to the next 1 bit (only if != IteratorEnd). */
+        constexpr Iterator& operator++() noexcept
+        {
+            m_val &= m_val - I{1U};
+            if (m_val != 0) m_pos = std::countr_zero(m_val);
+            return *this;
+        }
+        /** Get the current bit position (only if != IteratorEnd). */
+        constexpr const unsigned& operator*() const noexcept { return m_pos; }
+    };
+
+public:
+    /** Construct an all-zero bitset. */
+    constexpr IntBitSet() noexcept : m_val{0} {}
+    /** Copy construct a bitset. */
+    constexpr IntBitSet(const IntBitSet&) noexcept = default;
+    /** Construct from a list of values. */
+    constexpr IntBitSet(std::initializer_list<unsigned> ilist) noexcept : m_val(0)
+    {
+        for (auto i : ilist) m_val |= I{1U} << i;
+    }
+    /** Copy assign a bitset. */
+    constexpr IntBitSet& operator=(const IntBitSet&) noexcept = default;
+    /** Assign a list of values. */
+    constexpr IntBitSet& operator=(std::initializer_list<unsigned> ilist) noexcept
+    {
+        m_val = 0;
+        for (auto i : ilist) m_val |= I{1U} << i;
+        return *this;
+    }
+    /** Construct a bitset with the singleton i. */
+    static constexpr IntBitSet Singleton(unsigned i) noexcept {
+        return IntBitSet(I(1U) << i);
+    }
+    /** Construct a bitset with bits 0..count-1 (inclusive) set to 1. */
+    static constexpr IntBitSet Fill(unsigned count) noexcept {
+        IntBitSet ret;
+        if (count) ret.m_val = I(~I{0}) >> (MAX_SIZE - count);
+        return ret;
+    }
+    /** Compute the number of 1 bits in the bitset. */
+    constexpr unsigned Count() const noexcept { return PopCount(m_val); }
+    /** Return the number of bits that this object holds. */
+    static constexpr unsigned Size() noexcept { return MAX_SIZE; }
+    /** Set a bit to 1. */
+    constexpr void Set(unsigned pos) noexcept { m_val |= I{1U} << pos; }
```

Comment:
`std::bitset` throws for these if pos is out of range.

Maybe add `Assume()`s? Or were you trying to keep the header self-contained?